rfcs/0004-range-queries.md

```diff
@@ -0,0 +1,217 @@
+# SlateDB Range Scans
+
+<!-- TOC start (generated with https://github.com/derlin/bitdowntoc) -->
+
+- [Motivation](#motivation)
+- [Goals](#goals)
+- [Public API](#public-api)
+- [Implementation](#implementation)
+   * [Checkpoint References](#checkpoint-references)
+   * [Memory Usage](#memory-usage)
+- [Rejected Alternatives & Follow-Ups](#rejected-alternatives-follow-ups)
+- [Updates](#updates)
+
+<!-- TOC end -->
+
+Status: Discussion
+
+Authors:
+
+* [Jason Gustafson](https://github.com/hachikuji)
+
+References:
+
+* https://github.com/slatedb/slatedb/issues/8
+* https://github.com/facebook/rocksdb/wiki/Iterator
+* https://github.com/facebook/rocksdb/wiki/Iterator-Implementation
+* https://github.com/slatedb/slatedb/pull/260
+* https://github.com/slatedb/slatedb/pull/278
+
+## Motivation
+
+
+The ability to scan the keys in a range is an essential feature for building applications.
+For example, this can be used to implement a collections of keys using a common prefix. 
+Currently, SlateDB only supports fetching individual values for known keys.
+
+## Goals
+
+The goal of this work is to provide a flexible capability to query a range of keys in the
+database. It should be possible to specify unbounded ranges to allow querying all keys,
+or bounded ranges for finer precision and efficiency.
+
+Range queries have the potential to be expensive from a resource perspective because they
+hold references to existing tables, which prevents garbage collection. Our goal here is
+to provide a simple initial approach to state management which optimizes for common cases,
+while leaving room for further optimization in the future.
+
+## Public API
+
+We introduce a new `Db::scan` API which accept a `DbRange` defining the range of the scan.
+The result will be a `DbIterator` object which iterates the current records within the range.
+Users may either provide an explicit starting range in the call to `scan`, or they
+may `seek` using the returned iterator.
+
+```rust
+impl Db {
+    /// Scan a range of keys.
+    /// 
+    /// returns a `DbIterator`
+    pub async fn scan(
+        &self,
+        range: std::ops::Range<&[u8]>
+    ) -> Result<DbIterator, SlateDbError>;
+}
+
+impl DbIterator {
+    /// Get the next record in the scan.
+    /// 
+    /// returns Ok(None) when the scan is complete
+    /// returns Err(InvalidatedIterator) if the iterator has been invalidated
+    ///  due to an underlying error
+    pub async fn next(
```

Comment:
https://github.com/facebook/rocksdb/wiki/Iterator 

RocksDB iterator provides both `next` and `prev`.

Being able to iterate from first key to last key, as well as from last_key to first_key would enable more scenarios. It will be useful for the listed use case `collections of keys using a common prefix.` 

Would implementing `prev` add significantly more complexity?

Reply:
Agree with @vigneshc, going backward is a useful feature. In some cases it could be emulated by encoding the keys in a such way that they sort in reverse order, but that feels unnatural, also sometimes one might need both forward and backward order.

Reply:
it does add more complexity, but it might still be worthwhile. If it's ok with you, could we do this in a subsequent update to the RFC? By then I'll hopefully have a better understanding of the complexity it would require. Off the top of my head, the only issue I can imagine is the need to return records from a block in reverse order. I guess that would require either the ability to iterate backwards (not sure if possible), or potentially buffer all prior records in the block.

Reply:
the reverse iterator has proven useful for queries in MyRocks particularly for handling `ORDER BY xx ASC` or `DESC` clauses, when you got `DESC`, you can simple switch the iterator to the reverse one in RocksDB.

but for now, a single-direction iterator in slatedb can already provide significant benefits for many workloads. i guess the reverse iterator can be introduced as a self-contained feature after the single-direction iterator get implemented.

Reply:
yeah, separate item for reverse should be good. Shall we track it as a an issue and link to it in this RFC?